Added guardduty filter resource

graffzon opened the pull request:
<!--- See what makes a good Pull Request at : https://github.com/terraform-providers/terraform-provider-aws/blob/master/.github/CONTRIBUTING.md#pull-requests --->

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

<!--- Thank you for keeping this note for the community --->

<!--- If your PR fully resolves and should automatically close the linked issue, use Closes. Otherwise, use Relates --->
Relates OR Closes #0000 (Sorry, I didn't get, shall I create an issue for this PR or the PR on its own is enough?)

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):
<!--
If change is not user facing, just write "NONE" in the release-note block below.
-->

```release-note
Added a GuardDuty Filter resource.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSGuardDuty/Filter'
    --- PASS: TestAccAWSGuardDuty/Filter (34.36s)
        --- PASS: TestAccAWSGuardDuty/Filter/basic (16.44s)
        --- PASS: TestAccAWSGuardDuty/Filter/import (17.92s)
...
```

graffzon commented:
Hi there! I see you have hundreds of pull requests, folks :( Is there any chance I can speed up a review for this one?

gdavison commented on website/docs/r/guardduty_filter.html.markdown:
```diff
@@ -0,0 +1,75 @@
+---
+layout: "aws"
```
Our current documentation standards have added the frontmatter entry `subcategory: "GuardDuty"`

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,493 @@
+				Required: true,
+				ForceNew: true,
+			},
+			"name": {
```
Please add a validator for a minimum length of 3, maximum length of 64

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,493 @@
+				ForceNew: true,
+			},
+			"description": {
+				Type:     schema.TypeString,
```
Please add a validator for a maximum length of 512

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,493 @@
+		return err
+	}
+
+	tagsInterface := d.Get("tags").(map[string]interface{})
```
Please update this to use the new `keyvaluetags` package at https://github.com/terraform-providers/terraform-provider-aws/tree/master/aws/internal/keyvaluetags

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,493 @@
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Reading GuardDuty Filter '%s' failed: %s", name, err.Error())
```
Please update the `fmt.Errorf` calls to return the error object itself using the `%w` directive

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,493 @@
+			//  a combination of these filters.
+			value, err := conditionValueToInt(typedCriterion["values"].([]interface{}))
+			if err != nil {
+				return nil, fmt.Errorf("Value seems to be not an integer: %v", typedCriterion["values"].([]interface{})[0])
```
These error messages can be rephrased to `"Value must be an integer: %v"`

gdavison commented on aws/resource_aws_guardduty_filter_test.go:
```diff
@@ -0,0 +1,133 @@
+		},
+	})
+}
+
```
This resource should have some tests that cover updating values

gdavison commented on aws/resource_aws_guardduty_filter_test.go:
```diff
@@ -0,0 +1,133 @@
+				Config: testAccGuardDutyFilterConfig_full(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsGuardDutyFilterExists(resourceName),
+					resource.TestCheckResourceAttrSet(resourceName, "detector_id"),
```
Use `resource.TestCheckResourceAttrPair` to verify that the correct value is set

gdavison commented on aws/resource_aws_guardduty_filter_test.go:
```diff
@@ -0,0 +1,133 @@
+					resource.TestCheckResourceAttr(resourceName, "name", "test-filter"),
+					resource.TestCheckResourceAttr(resourceName, "action", "ARCHIVE"),
+					resource.TestCheckResourceAttr(resourceName, "rank", "1"),
+				),
```
Please add some tests for the `finding_criteria`

gdavison commented:
Hi @graffzon, are you still interested in working on this PR?

suzuki-shunsuke commented:
@graffzon @gdavison
Thank you for your great work.
I'm interested in this pull request, but this seems to be pending.

Can I contribute to this pull request?
If I can, how should I do?

I checked out this pull request and added some commits according to the review comments.

* https://github.com/terraform-providers/terraform-provider-aws/compare/master...suzuki-shunsuke:added-guardduty-filter-resource

Should I create a new pull request? Or is there any other way?

Thank you.

graffzon commented:
Hi folks, my apologies for keeping silence. @suzuki-shunsuke if you have time and willingness to contribute to this PR - I'd be happy. Unfortunately I don't have a prepared setup to test the GuardDuty resources anymore. However, if it will hang for a couple more months, I might have time to come back to it 😅

graffzon commented:
@suzuki-shunsuke I found a way to merge your changes into this PR, thanks a lot for your work.
@gdavison could you please take another look at the PR after the changes made by @suzuki-shunsuke ?

suzuki-shunsuke commented:
Thanks a lot.
I'll add commits to fix the failure of CI.

* errors
  * [awsproviderlint - AWSR002: missing (keyvaluetags.KeyValueTags).IgnoreConfig()](https://github.com/terraform-providers/terraform-provider-aws/pull/10676/checks?check_run_id=842275982)
  * [tfproviderdocs check -  YAML frontmatter should not contain sidebar_current](https://github.com/terraform-providers/terraform-provider-aws/pull/10676/checks?check_run_id=842278427)
* added commits
  * https://github.com/terraform-providers/terraform-provider-aws/compare/master...suzuki-shunsuke:added-guardduty-filter-resource
  * https://github.com/terraform-providers/terraform-provider-aws/commit/6d1a03f270fa998139f11314e12308da04e2e506
  * https://github.com/terraform-providers/terraform-provider-aws/commit/045794d9c08394cb4985f133f93425fb7614aa36
  * https://github.com/terraform-providers/terraform-provider-aws/commit/53d0e17f2d15de4b518e25d5237dae3ad45a0cc8

graffzon commented:
@suzuki-shunsuke Thanks for being that quick :)
I didn't manage to merge your PR because it contained an update from upstream and produced a weird changes for me, so I just updated my fork from the upstream and added the code you implemented in your commits, tagging you in the commits' description. By some reason it still fails, I will have a look later today

suzuki-shunsuke commented:
@graffzon 

> I didn't manage to merge your PR because it contained an update from upstream and produced a weird changes for me

Would you cherry pick 045794d and 53d0e17 instead of bb3819c 255ff95?

``` 
$ git remote add suzuki-shunsuke https://github.com/suzuki-shunsuke/terraform-provider-aws 
$ git remote -v
origin	https://github.com/graffzon/terraform-provider-aws (fetch)
origin	https://github.com/graffzon/terraform-provider-aws (push)
suzuki-shunsuke	https://github.com/suzuki-shunsuke/terraform-provider-aws (fetch)
suzuki-shunsuke	https://github.com/suzuki-shunsuke/terraform-provider-aws (push)

$ git fetch suzuki-shunsuke
$ git reset HEAD~2 --hard # remove bb3819c 255ff95
$ git cherry-pick 045794d and 53d0e17
$ git push origin added-guardduty-filter-resource --force
```

gdavison commented on website/docs/r/guardduty_filter.html.markdown:
```diff
@@ -0,0 +1,75 @@
+  name        = "MyFilter"
+  action      = "ARCHIVE"
+  description = "Some description"
+  detector_id = "123456271278c0df5e089123480d8765"
```
This could show a resource reference. Optionally, the `aws_guardduty_detector` could be included, but it's not necessary
```suggestion
  detector_id = aws_guardduty_detector.example.id
```

gdavison commented on website/docs/r/guardduty_filter.html.markdown:
```diff
@@ -0,0 +1,75 @@
+* `name` - (Required) The name of your filter.
+* `description` - (Optional) Description of the filter.
+* `rank` - (Required) Specifies the position of the filter in the list of current filters. Also specifies the order in which this filter is applied to the findings.
+* `action` - (Required) Specifies the action that is to be applied to the findings that match the filter. Can be one of (ARCHIVE|NOOP).
```
Styling nit

```suggestion
* `action` - (Required) Specifies the action that is to be applied to the findings that match the filter. Can be one of `ARCHIVE` or `NOOP`.
```

suzuki-shunsuke replied:
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/ccd2fdccffc9ad4bd0e65e1dbc26776415f75b67

gdavison commented on website/docs/r/guardduty_filter.html.markdown:
```diff
@@ -0,0 +1,75 @@
+* `rank` - (Required) Specifies the position of the filter in the list of current filters. Also specifies the order in which this filter is applied to the findings.
+* `action` - (Required) Specifies the action that is to be applied to the findings that match the filter. Can be one of (ARCHIVE|NOOP).
+* `tags` (Optional) - The tags that you want to add to the Filter resource. A tag consists of a key and a value.
+* `finding_criteria` (Required) - Represents the criteria to be used in the filter for querying findings. A list, consists of `criterion` structures, each having required `condition`, `field` and `values` fields. See the example for the structure and see the [AWS Documentation](https://docs.aws.amazon.com/guardduty/latest/ug/create-filter.html) for the list of available fields for the `field` att ...
```
The `finding_criteria` block should be described in a separate section. See, for example, the `aws_kinesis_analytics_application` documentation

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+
+	criteria := make(map[string]*schema.Set)
+
+	criteria["criterion"] = schema.NewSet(schema.HashResource(criterionResource()), flatCriteria)
```
`criterionResource()` should not redefine the schema here. The best option is probably to define a hash function that can be used here and set it on the schema as well.

suzuki-shunsuke replied:
https://github.com/terraform-providers/terraform-provider-aws/commit/bea0cb4a926671cf1a51cf349bc02839b533360c

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+	flatCriterion["field"] = field
+	flatCriterion["condition"] = conditionName
+	flatCriterion["values"] = make([]interface{}, 1)
+	flatCriterion["values"].([]interface{})[0] = strconv.FormatInt(*conditionValue, 10)
```
`strconv.Itoa()` would be a simpler function to use here, since we don't need to specify the base.

suzuki-shunsuke replied:
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/ee392bd4d8555f2f45a87f9ddc58b04cfe327803

suzuki-shunsuke replied:
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/ee392bd4d8555f2f45a87f9ddc58b04cfe327803

suzuki-shunsuke replied:
The type of `*conditionValue` is `int64` but the type of the argument `strconv.Itoa` is `int`.
So `strconv.FormatInt` isn't wrong.

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+
+	for field, conditions := range findingCriteriaRemote.Criterion {
+		if len(conditions.Equals) > 0 {
+			flatCriteria = append(flatCriteria, flattenStringCondition(field, "equals", conditions.Equals))
```
The two `flattenStringCondition()` calls here could take a `[]string` parameter, and have the dereference happen here
```suggestion
			flatCriteria = append(flatCriteria, flattenStringCondition(field, "equals", aws.StringValueSlice(conditions.Equals)))
```

suzuki-shunsuke replied:
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/59db7dd4624918589b69266026dd266bf44fa71a

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+			flatCriteria = append(flatCriteria, flattenStringCondition(field, "not_equals", conditions.NotEquals))
+		}
+		if conditions.GreaterThan != nil {
+			flatCriteria = append(flatCriteria, flattenIntCondition(field, "greater_than", conditions.GreaterThan))
```
The four `flattenIntCondition()` calls here could take an `int` parameter and have the dereference happen here
```suggestion
			flatCriteria = append(flatCriteria, flattenIntCondition(field, "greater_than", int(aws.Int64Value(conditions.GreaterThan))))
```

suzuki-shunsuke replied:
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/59db7dd4624918589b69266026dd266bf44fa71a

gdavison commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: validation.StringInSlice([]string{
+					"NOOP",
+					"ARCHIVE",
```
These should be replaced by the constants defined in the AWS SDK
```suggestion
					guardduty.FilterActionNoop,
					guardduty..FilterActionArchive,
```

suzuki-shunsuke replied:
https://github.com/terraform-providers/terraform-provider-aws/commit/5d4dd0807276dabbfa23c926293cd4bfcf35da56

suzuki-shunsuke commented:
@gdavison Thank you for your review.
I'll deal with it.

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+				Required: true,
+			},
+		},
+		Timeouts: &schema.ResourceTimeout{
+			Create: schema.DefaultTimeout(60 * time.Second),
+			Update: schema.DefaultTimeout(60 * time.Second),
+		},
```
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/f12249e93f159d046e39b50e47dfdd23c1fd2638

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+	log.Printf("[DEBUG] Creating GuardDuty Filter: %s", input)
+	output, err := conn.CreateFilter(&input)
+	if err != nil {
+		return fmt.Errorf("Creating GuardDuty Filter %s failed: %s", input, err.Error())
```
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/76b1c4c8f7a08069b8ccd8ba4c183ca695a977be

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+
+	_, err = conn.UpdateFilter(&input)
+	if err != nil {
+		return fmt.Errorf("Updating GuardDuty Filter with ID %s failed: %s", d.Id(), err.Error())
```
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/b8e03f3ee7358d59a716c1935eac3afa165ab370

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+
+	_, err := conn.DeleteFilter(&input)
+	if err != nil {
+		return fmt.Errorf("Deleting GuardDuty Filter '%s' failed: %s", d.Id(), err.Error())
```
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/3d0707dbbe3a159498d1365ea7cc233ad90e941e

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+
+func parseImportedId(importedId string) (string, string, error) {
+	parts := strings.SplitN(importedId, "_", 2)
+	if len(parts) < 2 {
```
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/9710499560d5b8ad7cf1c1cc9b88588842e72f6b

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter_test.go:
```diff
@@ -0,0 +1,183 @@
+
+func testAccGuardDutyFilterConfig_full() string {
+	return fmt.Sprintf(`
+%[1]s
+
+resource "aws_guardduty_filter" "test" {
+  detector_id = "${aws_guardduty_detector.test.id}"
+	name        = "test-filter"
+	action      = "ARCHIVE"
+	rank        = 1
+
+  finding_criteria {
+    criterion {
+      field     = "region"
+      values    = ["eu-west-1"]
+      condition = "equals"
+    }
+
+    criterion {
+      field     = "service.additionalInfo.threatListName"
+      values    = ["some-threat", "another-threat"]
+      condition = "not_equals"
+    }
+
+    criterion {
+      field     = "updatedAt"
+      values    = ["1570744740000"]
+      condition = "less_than"
+    }
+
+    criterion {
+      field     = "updatedAt"
+      values    = ["1570744240000"]
+      condition = "greater_than"
+    }
+  }
+}`, testAccGuardDutyDetectorConfig_basic3)
```
https://github.com/suzuki-shunsuke/terraform-provider-aws/commit/5db748b1191ffef99dea4ed99ff5966a95c19bcc

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter_test.go:
```diff
@@ -0,0 +1,183 @@
+			continue
+		}
+
+		parts := strings.SplitN(rs.Primary.ID, "_", 2)
```
https://github.com/terraform-providers/terraform-provider-aws/commit/7ab6e040af931af5cd871d2559c6234d7b55acfb

suzuki-shunsuke replied:
https://github.com/terraform-providers/terraform-provider-aws/commit/a17806d8784760c5495854e1dbb4b446cc5ab470

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+		return fmt.Errorf("Creating GuardDuty Filter %s failed: %s", input, err.Error())
+	}
+
+	d.SetId(strings.Join([]string{d.Get("detector_id").(string), *output.Name}, "_"))
```
https://github.com/terraform-providers/terraform-provider-aws/commit/cd8003316bf3c037f089a3563bed7833ac086a9a

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter.go:
```diff
@@ -0,0 +1,489 @@
+		return fmt.Errorf("Creating GuardDuty Filter %s failed: %s", input, err.Error())
+	}
+
+	d.SetId(strings.Join([]string{d.Get("detector_id").(string), *output.Name}, "_"))
```
https://github.com/terraform-providers/terraform-provider-aws/commit/f4aa17d3f738bcfa7f4b2244eb63c828fbb04feb

suzuki-shunsuke commented on aws/resource_aws_guardduty_filter_test.go:
```diff
@@ -0,0 +1,183 @@
+				Config: testAccGuardDutyFilterConfig_full(),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckAwsGuardDutyFilterExists(resourceName),
+					resource.TestCheckResourceAttrPair(resourceName, "detector_id", "aws_guardduty_detector.test", "id"),
```
https://github.com/terraform-providers/terraform-provider-aws/commit/e46821ecb19d6578aa161b8e69c9e6842e930326

gdavison commented:
Hi @graffzon and @suzuki-shunsuke. We'd like to get this merged in. @graffzon do you want to merge @suzuki-shunsuke's changes into your PR, or @suzuki-shunsuke do you want to create a PR from your branch?

We'll also need to update the version of the Terraform SDK plugin to v2. Update references to the package `github.com/hashicorp/terraform-plugin-sdk/*` to `github.com/hashicorp/terraform-plugin-sdk/v2/*`.

graffzon commented:
Thanks for your tremendous effort @gdavison and @suzuki-shunsuke !!! 🙏 
I merged the branch from @suzuki-shunsuke and rebased from the latest master.
Just not sure why it says only 4 checks when I'm pretty sure I saw smth like 12 before rebasing. I'm not very familiar with the Github Checks, it's a rather new feature that I haven't used yet. Might it have just skipped some checks? Like I don't actually see the tests being ran e.g.
And about the review, sorry I've dropped out from the review process, is everything incorporated now?

gdavison commented:
Thanks again for all the work you've put in, @graffzon and @suzuki-shunsuke. There are a few remaining changes we'd like to see before merging the PR.

If one of you would like to make the changes go ahead. Otherwise, I can make the changes next week (24 August) so that we can get it in for the release that week.
